File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <deque>
#include <algorithm>
#include <vector>
#include <iostream>

using namespace std;

vector<vector<string>> get_subvector(vector<string> ticket, vector<vector<string>> tickets){
    vector<vector<string>> subvector = tickets;
    subvector.erase(find(subvector.begin(), subvector.end(), ticket));
    return subvector;
}

vector<string> dfs(string airport, vector<vector<string>> tickets){
    vector<string> path;
    if (tickets.size() == 0){
        path.push_back(airport);
        return path;
    }
    
    vector<vector<string>> path_list;
    for (vector<string> ticket: tickets){
        if (ticket[0] == airport){
            vector<string> tmp_path = dfs(ticket[1], get_subvector(ticket, tickets));
            if (tmp_path.size() > 0){
                path_list.push_back(tmp_path);
            }
        }
    }
    if (path_list.size() > 0){
        sort(path_list.begin(), path_list.end());
        path = path_list.front();
        path.insert(path.begin(), airport);
    }
    return path;
}

vector<string> solution(vector<vector<string>> tickets) {
    return dfs("ICN", tickets);
}

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public String[] solution(String[][] tickets) {
        return searchPath("ICN", List.of(tickets)).toArray(new String[0]);
    }

    public List<String> searchPath(String currentCity, List<String[]> tickets){
        if (tickets.size() == 0){
            List<String> path = new ArrayList<>();
            path.add(currentCity);
            return path;
        }

        List<String[]> nextTickets = getNextTickets(currentCity, tickets);
        List<List<String>> pathList = new ArrayList<>();

        for (String[] ticket: nextTickets){
            List<String[]> remainTickets = new ArrayList<>(tickets);
            remainTickets.remove(ticket);

            List<String> path = searchPath(ticket[1], remainTickets);
            if (path != null){
                path.add(0, currentCity);
                pathList.add(path);
            }
        }

        //pathList 알파벳 순으로 정렬
        pathList.sort((o1, o2) -> {
            for (int i = 0; i < o1.size(); i++){
                if (o1.get(i).compareTo(o2.get(i)) != 0){
                    return o1.get(i).compareTo(o2.get(i));
                }
            }
            return 0;
        });
        if (pathList.size() > 0){
            return pathList.get(0);
        } else {
            return null;
        }
    }

    public List<String[]> getNextTickets(String currentCity, List<String[]> remainTickets){
        List<String[]> nextTickets = new ArrayList<>();

        for(String[] ticket : remainTickets){
            if(ticket[0].equals(currentCity)){
                nextTickets.add(ticket);
            }
        }

        return nextTickets;
    }
}